{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2e28c71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting powerlaw\n",
      "  Downloading powerlaw-1.5-py3-none-any.whl (24 kB)\n",
      "Requirement already satisfied: mpmath in /home/maxim/.local/lib/python3.8/site-packages (from powerlaw) (1.2.1)\n",
      "Requirement already satisfied: matplotlib in /usr/lib/python3/dist-packages (from powerlaw) (3.1.2)\n",
      "Requirement already satisfied: scipy in /home/maxim/.local/lib/python3.8/site-packages (from powerlaw) (1.10.1)\n",
      "Requirement already satisfied: numpy in /home/maxim/.local/lib/python3.8/site-packages (from powerlaw) (1.21.2)\n",
      "Installing collected packages: powerlaw\n",
      "Successfully installed powerlaw-1.5\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install powerlaw\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3799a2a4",
   "metadata": {},
   "source": [
    "A trusted reference sequence is to be inserted into the \"RefSeq\" folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "59cc546e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/absolute/path/to/Documents/Numbering_Scheme/RefSeq/seq_ref_tem-1.fasta\n"
     ]
    }
   ],
   "source": [
    "seq_file_path = \"/absolute/path/to/Documents/Numbering_Scheme/RefSeq/seq_ref_tem-1.fasta\"\n",
    "print(seq_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "345c67a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmsearch :: search profile(s) against a sequence database\n",
      "# HMMER 3.3 (Nov 2019); http://hmmer.org/\n",
      "# Copyright (C) 2019 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# query HMM file:                  ./Documents/HMM/Profile_HMM.hmm\n",
      "# target sequence database:        ./Documents/Numbering_Scheme/RefSeq/seq_ref_tem_1.fasta\n",
      "# output directed to file:         ./Documents/Numbering_Scheme/numbering/TEM-1_numbering.txt\n",
      "# sequence reporting threshold:    E-value <= 0.01\n",
      "# domain reporting threshold:      E-value <= 0.01\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "Query:       Profile_HMM  [M=286]\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                                             Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                                             -----------\n",
      "   1.5e-191  620.9   0.1   1.7e-191  620.7   0.1    1.0  1  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562|  \n",
      "\n",
      "\n",
      "Domain annotation for each sequence (and alignments):\n",
      ">> sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562|  \n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  620.7   0.1  1.7e-191  1.7e-191      24     286 .]       1     263 []       1     263 [] 1.00\n",
      "\n",
      "  Alignments for each domain:\n",
      "  == domain 1  score: 620.7 bits;  conditional E-value: 1.7e-191\n",
      "                                           Profile_HMM  24 hpetlvkvkdaedklgarvgyieldlnsgkilesfrpeerfpmlstfkvllcgavls 80 \n",
      "                                                           hpetlvkvkdaed+lgarvgyieldlnsgkilesfrpeerfpm+stfkvllcgavls\n",
      "  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562|   1 HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMSTFKVLLCGAVLS 57 \n",
      "                                                           9******************************************************** PP\n",
      "\n",
      "                                           Profile_HMM  81 rvdagqeqlgrrihysqndlveyspvtekhltdgmtvrelcsaaitmsdntaanlll 137\n",
      "                                                           rvdagqeqlgrrihysqndlveyspvtekhltdgmtvrelcsaaitmsdntaanlll\n",
      "  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562|  58 RVDAGQEQLGRRIHYSQNDLVEYSPVTEKHLTDGMTVRELCSAAITMSDNTAANLLL 114\n",
      "                                                           ********************************************************* PP\n",
      "\n",
      "                                           Profile_HMM 138 ttiggpkeltaflhnmgdhvtrldrwepelneaipnderdttmpaamattlrklltg 194\n",
      "                                                           ttiggpkeltaflhnmgdhvtrldrwepelneaipnderdttmpaamattlrklltg\n",
      "  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562| 115 TTIGGPKELTAFLHNMGDHVTRLDRWEPELNEAIPNDERDTTMPAAMATTLRKLLTG 171\n",
      "                                                           ********************************************************* PP\n",
      "\n",
      "                                           Profile_HMM 195 elltlasrqqlidwmeadkvagpllrsalpagwfiadksgagergsrgiiaalgpdg 251\n",
      "                                                           elltlasrqqlidwmeadkvagpllrsalpagwfiadksgagergsrgiiaalgpdg\n",
      "  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562| 172 ELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGAGERGSRGIIAALGPDG 228\n",
      "                                                           ********************************************************* PP\n",
      "\n",
      "                                           Profile_HMM 252 kpsrivviyttgsqatmdernrqiaeigaslikhw 286\n",
      "                                                           kpsrivviyttgsqatmdernrqiaeigaslikhw\n",
      "  sid|842|pid|818|hfid|1|sfid|1|gi|46015634|taxid|562| 229 KPSRIVVIYTTGSQATMDERNRQIAEIGASLIKHW 263\n",
      "                                                           *********************************** PP\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (286 nodes)\n",
      "Target sequences:                          1  (263 residues searched)\n",
      "Passed MSV filter:                         1  (1); expected 0.0 (0.02)\n",
      "Passed bias filter:                        1  (1); expected 0.0 (0.02)\n",
      "Passed Vit filter:                         1  (1); expected 0.0 (0.001)\n",
      "Passed Fwd filter:                         1  (1); expected 0.0 (1e-05)\n",
      "Initial search space (Z):                  1  [actual number of targets]\n",
      "Domain search space  (domZ):               1  [number of targets reported over threshold]\n",
      "# CPU time: 0.01u 0.01s 00:00:00.02 Elapsed: 00:00:00.03\n",
      "# Mc/sec: 1.98\n",
      "//\n",
      "[ok]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Run HMMER to search for hits in your sequence, Profile_HMM.hmm file will be aligned to\n",
    "# seq_ref_tem_1.fasta file, out of this alignment the TEM-1_numbering.txt file will be created.\n",
    "hmmsearch_cmd = ['hmmsearch', '-E', '0.01', '--domE', '0.01', \n",
    "                 '-o', './Documents/Numbering_Scheme/numbering/TEM-1_numbering.txt',\n",
    "                 './Documents/HMM/Profile_HMM.hmm',\n",
    "                 './Documents/Numbering_Scheme/RefSeq/seq_ref_tem_1.fasta']\n",
    "subprocess.run(hmmsearch_cmd, check=True)\n",
    "\n",
    "# Read in the output file and print it out\n",
    "with open('./Documents/Numbering_Scheme/numbering/TEM-1_numbering.txt', 'r') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5d042539",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     amino_acid_number Profile_HMM\n",
      "0                    1           H\n",
      "1                    2           P\n",
      "2                    3           E\n",
      "3                    4           T\n",
      "4                    5           L\n",
      "5                    6           V\n",
      "6                    7           K\n",
      "7                    8           V\n",
      "8                    9           K\n",
      "9                   10           D\n",
      "10                  11           A\n",
      "11                  12           E\n",
      "12                  13           D\n",
      "13                  14           K\n",
      "14                  15           L\n",
      "15                  16           G\n",
      "16                  17           A\n",
      "17                  18           R\n",
      "18                  19           V\n",
      "19                  20           G\n",
      "20                  21           Y\n",
      "21                  22           I\n",
      "22                  23           E\n",
      "23                  24           L\n",
      "24                  25           D\n",
      "25                  26           L\n",
      "26                  27           N\n",
      "27                  28           S\n",
      "28                  29           G\n",
      "29                  30           K\n",
      "30                  31           I\n",
      "31                  32           L\n",
      "32                  33           E\n",
      "33                  34           S\n",
      "34                  35           F\n",
      "35                  36           R\n",
      "36                  37           P\n",
      "37                  38           E\n",
      "38                  39           E\n",
      "39                  40           R\n",
      "40                  41           F\n",
      "41                  42           P\n",
      "42                  43           M\n",
      "43                  44           L\n",
      "44                  45           S\n",
      "45                  46           T\n",
      "46                  47           F\n",
      "47                  48           K\n",
      "48                  49           V\n",
      "49                  50           L\n",
      "50                  51           L\n",
      "51                  52           C\n",
      "52                  53           G\n",
      "53                  54           A\n",
      "54                  55           V\n",
      "55                  56           L\n",
      "56                  57           S\n",
      "57                  58           R\n",
      "58                  59           V\n",
      "59                  60           D\n",
      "60                  61           A\n",
      "61                  62           G\n",
      "62                  63           Q\n",
      "63                  64           E\n",
      "64                  65           Q\n",
      "65                  66           L\n",
      "66                  67           G\n",
      "67                  68           R\n",
      "68                  69           R\n",
      "69                  70           I\n",
      "70                  71           H\n",
      "71                  72           Y\n",
      "72                  73           S\n",
      "73                  74           Q\n",
      "74                  75           N\n",
      "75                  76           D\n",
      "76                  77           L\n",
      "77                  78           V\n",
      "78                  79           E\n",
      "79                  80           Y\n",
      "80                  81           S\n",
      "81                  82           P\n",
      "82                  83           V\n",
      "83                  84           T\n",
      "84                  85           E\n",
      "85                  86           K\n",
      "86                  87           H\n",
      "87                  88           L\n",
      "88                  89           T\n",
      "89                  90           D\n",
      "90                  91           G\n",
      "91                  92           M\n",
      "92                  93           T\n",
      "93                  94           V\n",
      "94                  95           R\n",
      "95                  96           E\n",
      "96                  97           L\n",
      "97                  98           C\n",
      "98                  99           S\n",
      "99                 100           A\n",
      "100                101           A\n",
      "101                102           I\n",
      "102                103           T\n",
      "103                104           M\n",
      "104                105           S\n",
      "105                106           D\n",
      "106                107           N\n",
      "107                108           T\n",
      "108                109           A\n",
      "109                110           A\n",
      "110                111           N\n",
      "111                112           L\n",
      "112                113           L\n",
      "113                114           L\n",
      "114                115           T\n",
      "115                116           T\n",
      "116                117           I\n",
      "117                118           G\n",
      "118                119           G\n",
      "119                120           P\n",
      "120                121           K\n",
      "121                122           E\n",
      "122                123           L\n",
      "123                124           T\n",
      "124                125           A\n",
      "125                126           F\n",
      "126                127           L\n",
      "127                128           H\n",
      "128                129           N\n",
      "129                130           M\n",
      "130                131           G\n",
      "131                132           D\n",
      "132                133           H\n",
      "133                134           V\n",
      "134                135           T\n",
      "135                136           R\n",
      "136                137           L\n",
      "137                138           D\n",
      "138                139           R\n",
      "139                140           W\n",
      "140                141           E\n",
      "141                142           P\n",
      "142                143           E\n",
      "143                144           L\n",
      "144                145           N\n",
      "145                146           E\n",
      "146                147           A\n",
      "147                148           I\n",
      "148                149           P\n",
      "149                150           N\n",
      "150                151           D\n",
      "151                152           E\n",
      "152                153           R\n",
      "153                154           D\n",
      "154                155           T\n",
      "155                156           T\n",
      "156                157           M\n",
      "157                158           P\n",
      "158                159           A\n",
      "159                160           A\n",
      "160                161           M\n",
      "161                162           A\n",
      "162                163           T\n",
      "163                164           T\n",
      "164                165           L\n",
      "165                166           R\n",
      "166                167           K\n",
      "167                168           L\n",
      "168                169           L\n",
      "169                170           T\n",
      "170                171           G\n",
      "171                172           E\n",
      "172                173           L\n",
      "173                174           L\n",
      "174                175           T\n",
      "175                176           L\n",
      "176                177           A\n",
      "177                178           S\n",
      "178                179           R\n",
      "179                180           Q\n",
      "180                181           Q\n",
      "181                182           L\n",
      "182                183           I\n",
      "183                184           D\n",
      "184                185           W\n",
      "185                186           M\n",
      "186                187           E\n",
      "187                188           A\n",
      "188                189           D\n",
      "189                190           K\n",
      "190                191           V\n",
      "191                192           A\n",
      "192                193           G\n",
      "193                194           P\n",
      "194                195           L\n",
      "195                196           L\n",
      "196                197           R\n",
      "197                198           S\n",
      "198                199           A\n",
      "199                200           L\n",
      "200                201           P\n",
      "201                202           A\n",
      "202                203           G\n",
      "203                204           W\n",
      "204                205           F\n",
      "205                206           I\n",
      "206                207           A\n",
      "207                208           D\n",
      "208                209           K\n",
      "209                210           S\n",
      "210                211           G\n",
      "211                212           A\n",
      "212                213           G\n",
      "213                214           E\n",
      "214                215           R\n",
      "215                216           G\n",
      "216                217           S\n",
      "217                218           R\n",
      "218                219           G\n",
      "219                220           I\n",
      "220                221           I\n",
      "221                222           A\n",
      "222                223           A\n",
      "223                224           L\n",
      "224                225           G\n",
      "225                226           P\n",
      "226                227           D\n",
      "227                228           G\n",
      "228                229           K\n",
      "229                230           P\n",
      "230                231           S\n",
      "231                232           R\n",
      "232                233           I\n",
      "233                234           V\n",
      "234                235           V\n",
      "235                236           I\n",
      "236                237           Y\n",
      "237                238           T\n",
      "238                239           T\n",
      "239                240           G\n",
      "240                241           S\n",
      "241                242           Q\n",
      "242                243           A\n",
      "243                244           T\n",
      "244                245           M\n",
      "245                246           D\n",
      "246                247           E\n",
      "247                248           R\n",
      "248                249           N\n",
      "249                250           R\n",
      "250                251           Q\n",
      "251                252           I\n",
      "252                253           A\n",
      "253                254           E\n",
      "254                255           I\n",
      "255                256           G\n",
      "256                257           A\n",
      "257                258           S\n",
      "258                259           L\n",
      "259                260           I\n",
      "260                261           K\n",
      "261                262           H\n",
      "262                263           W\n"
     ]
    }
   ],
   "source": [
    "from modules.number_extractor import extract_profile_hmm\n",
    "\n",
    "# The numbering will be bounded to the profile HMM, the count will be always from 1-263 \n",
    "# amino acids in this case\n",
    "\n",
    "input_file = './Documents/Numbering_Scheme/numbering/TEM-1_numbering.txt'\n",
    "output_file = './Documents/Numbering_Scheme/numbering/amino_acid_number.csv'\n",
    "prefix_str = \"Profile_HMM\"\n",
    "\n",
    "df = extract_profile_hmm(input_file, output_file, prefix_str)\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "696cb2a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from modules.Fastamerge import merge_fasta_files\n",
    "\n",
    "input_dir = \"./test/testnetwork/\"\n",
    "output_file = \"./Documents/TEM-merged.fasta\"\n",
    "\n",
    "merge_fasta_files(input_dir, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20defa22",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/maxim/.local/lib/python3.8/site-packages/pandas/core/computation/expressions.py:20: UserWarning: Pandas requires version '2.7.3' or newer of 'numexpr' (version '2.7.1' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n"
     ]
    }
   ],
   "source": [
    "from modules.fasta_parser import parse_fasta\n",
    "import pandas as pd\n",
    "\n",
    "# Uploade all of the query sequences\n",
    "\n",
    "fasta_file = './test/non_identical.fasta'\n",
    "parse_fa = parse_fasta(fasta_file)\n",
    "parse_fa = parse_fa.reset_index(drop=True)\n",
    "\n",
    "#print(parse_fa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "30c94210",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          name                                amino_acid_sequence\n",
      "0      TEM-195  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "1      TEM-163  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "2        TEM-9  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "3       TEM-39  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMLST...\n",
      "4      TEM-211  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "...        ...                                                ...\n",
      "5746  ANG15461  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "5747  ANG27513  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "5748  ANG27339  HPETLVKVKDAEDQLGARVGYIELDLNSGKVLESFRPEERFPMMST...\n",
      "5749  ANG31338  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "5750  ANG12435  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...\n",
      "\n",
      "[5751 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Insperct and remove manually all signal peptides\n",
    "\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFFAAFCFPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('DPQHFRVALIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFLAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVARRPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIKHFRVALIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVSLIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('QHFRVALIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALFPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFFAAFCLHVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVTLIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVTLIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFFAAFCLPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQHFRVALIPFFAAFCIPVFA', '')\n",
    "parse_fa['amino_acid_sequence'] = parse_fa['amino_acid_sequence'].str.replace('MSIQQYFRVALIPFFAAFCLPVFA', '')\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Find the indices of the rows with the specified names\n",
    "names_to_keep = ['TEM-1', 'TEM-116', 'TEM-171', 'TEM-181']\n",
    "keep_indices = parse_fa[parse_fa['name'].isin(names_to_keep)].index\n",
    "\n",
    "# Mark duplicates in the 'amino_acid_sequence' column, but keep the ones with the specified names\n",
    "duplicates_mask = parse_fa.duplicated(subset='amino_acid_sequence', keep=False)\n",
    "duplicates_mask[keep_indices] = False\n",
    "\n",
    "# Apply the mask to the original DataFrame\n",
    "unique_df = parse_fa[~duplicates_mask]\n",
    "unique_df = unique_df.reset_index(drop=True)\n",
    "print(unique_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9033c490",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import tempfile\n",
    "import subprocess\n",
    "\n",
    "# aligne all query sequences to the profile HMM \n",
    "\n",
    "for index, row in unique_df.iterrows():\n",
    "    # create a temporary file for the sequence in FASTA format\n",
    "    with tempfile.NamedTemporaryFile(mode='w', delete=False) as tmpfile:\n",
    "        tmpfile.write(f'>seq{index}\\n{row[\"amino_acid_sequence\"]}\\n')\n",
    "        tmpfile.flush()\n",
    "\n",
    "        # run hmmsearch on the temporary file\n",
    "        hmmsearch_cmd = ['hmmsearch', '-E', '0.01', '--domE', '0.01', '-o',\n",
    "                         f'./Documents/Numbering_Scheme/numbering/TEM-sequences_{index}.txt',\n",
    "                         './Documents/HMM/Profile_HMM.hmm',tmpfile.name]\n",
    "        result = subprocess.run(hmmsearch_cmd, capture_output=True, check=True)\n",
    "        print(result.stdout.decode())\n",
    "\n",
    "    # print the output file\n",
    "    # print the output file\n",
    "    with open(f'./Documents/Numbering_Scheme/numbering/TEM-sequences_{index}.txt', 'r') as f:\n",
    "        print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "047df67f",
   "metadata": {},
   "source": [
    "This entire next step, is used to extract using regex all \"start\" and \"end\" positions \n",
    "including the sequences. It uses X and Y coordinates in a way to follow the amino acids\n",
    "and the numbering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c571a481",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "# create an empty DataFrame to store the sequences\n",
    "df_sequences = pd.DataFrame(columns=[\"seq_id\", \"start_pos\", \"sequence\", \"end_pos\"])\n",
    "\n",
    "# loop over files in the current directory that match the naming pattern\n",
    "filepaths = glob.glob(\"./Documents/Numbering_Scheme/numbering/TEM-sequences_*.txt\")\n",
    "filepaths = sorted(filepaths, key=lambda x: int(re.findall(r'\\d+', x)[0]))\n",
    "\n",
    "# loop over the filepaths\n",
    "for i, filepath in enumerate(filepaths):\n",
    "    # extract the index from the filename\n",
    "    index = int(re.search(r\"TEM-sequences_(\\d+)\\.txt\", filepath).group(1))\n",
    "    # open the file and read its contents\n",
    "    with open(filepath, \"r\") as f:\n",
    "        contents = f.read()\n",
    "    # find matches using a regular expression pattern\n",
    "    pattern = r\"(Profile_HMM)\\s+(\\d+)\\s+([\\w\\.]+)\\s+(\\d+)\"\n",
    "    matches = re.findall(pattern, contents)\n",
    "    \n",
    "    #print(f\"File {i}: {filepath}\")\n",
    "    #print(f\"Matches: {matches}\")\n",
    "\n",
    "    # store the sequences in a list\n",
    "    sequences = []\n",
    "    for match in matches:\n",
    "        if match[0] == \"Profile_HMM\":\n",
    "            seq_id = match[0]\n",
    "            start_pos = int(match[1])\n",
    "            seq = match[2].replace(\".\", \"\").upper()\n",
    "            end_pos = int(match[3])\n",
    "            sequences.append((seq_id, start_pos, seq, end_pos))\n",
    "\n",
    "    # convert the list of sequences to a DataFrame and add it to the main DataFrame\n",
    "    df = pd.DataFrame(sequences, columns=[\"seq_id\", \"start_pos\", \"sequence\", \"end_pos\"])\n",
    "    # set the index for this sequence to start at 0\n",
    "    df.index -= df.index.min()\n",
    "    # add the sequence to the main DataFrame\n",
    "    df_sequences = pd.concat([df_sequences, df], ignore_index=True)\n",
    "\n",
    "# display the resulting DataFrame\n",
    "#print(df_sequences)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d0166616",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import re\n",
    "\n",
    "# create an empty DataFrame to store the sequences\n",
    "df_seq = pd.DataFrame(columns=[\"seq_id\", \"start_pos\", \"sequence\", \"end_pos\"])\n",
    "\n",
    "# loop over files in the current directory that match the naming pattern\n",
    "filepaths = glob.glob(\"./Documents/Numbering_Scheme/numbering/TEM-sequences_*.txt\")\n",
    "filepaths = sorted(filepaths, key=lambda x: int(re.findall(r'\\d+', x)[0]))\n",
    "\n",
    "# loop over the filepaths\n",
    "for i, filepath in enumerate(filepaths):\n",
    "    # extract the index from the filename\n",
    "    index = int(re.search(r\"TEM-sequences_(\\d+)\\.txt\", filepath).group(1))\n",
    "    # open the file and read its contents\n",
    "    with open(filepath, \"r\") as f:\n",
    "        contents = f.read()\n",
    "    # find matches using a regular expression pattern\n",
    "    pattern = r\"(seq\\d+|Profile_HMM)\\s+(\\d+)\\s+([\\w\\.]+)\\s+(\\d+)\"\n",
    "    matches = re.findall(pattern, contents)\n",
    "\n",
    "    # store the sequences in a list\n",
    "    sequences = []\n",
    "    for match in matches:\n",
    "        if match[0].startswith(\"seq\"):\n",
    "            seq_id = int(match[0].replace(\"seq\", \"\"))\n",
    "            start_pos = int(match[1])\n",
    "            seq = match[2].replace(\".\", \"\").upper()\n",
    "            end_pos = int(match[3])\n",
    "            sequences.append((seq_id, start_pos, seq, end_pos))\n",
    "\n",
    "    # convert the list of sequences to a DataFrame and add it to the main DataFrame\n",
    "    df = pd.DataFrame(sequences, columns=[\"seq_id\", \"start_pos\", \"sequence\", \"end_pos\"])\n",
    "    # set the index for this sequence to start at 0\n",
    "    df.index -= df.index.min()\n",
    "    # add the sequence to the main DataFrame\n",
    "    df_seq = pd.concat([df_seq, df], ignore_index=True)\n",
    "\n",
    "# save the resulting DataFrame to a CSV file\n",
    "#df_seq.to_csv(\"sequences.csv\", index=False)\n",
    "\n",
    "# display the resulting DataFrame\n",
    "#print(df_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58b3bf47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combined_prenum = pd.merge(df_seq, df_sequences, left_index=True, right_index=True)\n",
    "#print(df_combined_prenum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5bedd5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_x = df_combined_prenum[['seq_id_x', 'start_pos_x', 'sequence_x', 'end_pos_x']]\n",
    "df_y = df_combined_prenum[['seq_id_y', 'start_pos_y', 'sequence_y', 'end_pos_y']]\n",
    "#print(df_x)\n",
    "#print(df_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f3271669",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combined_prenum1 = df_combined_prenum.drop(['sequence_x', 'sequence_y'], axis=1)\n",
    "#print(df_combined_prenum1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "46ed5b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_grouped = df_combined_prenum.groupby(['seq_id_x', 'seq_id_y']).agg({\n",
    "    'start_pos_x': 'first', \n",
    "    'end_pos_x': 'last', \n",
    "    'start_pos_y': 'first', \n",
    "    'end_pos_y': 'last'\n",
    "}).reset_index()\n",
    "\n",
    "#print(df_grouped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "658b1db3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_start_end_x = df_grouped.loc[:, ['seq_id_x', 'start_pos_x', 'end_pos_x']]\n",
    "#print(df_start_end_x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "12485ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_start_end_y = df_grouped.loc[:, ['seq_id_y', 'start_pos_y', 'end_pos_y']]\n",
    "#print(df_start_end_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "15dd27e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define df_start_end_x and df_x here\n",
    "\n",
    "collapsed_seq_x = df_x.groupby('seq_id_x')['sequence_x'].agg(''.join).reset_index()\n",
    "merged_df_x = pd.merge(df_start_end_x, collapsed_seq_x, on='seq_id_x')\n",
    "\n",
    "# Replace the first row of sequence_x column with first_sequence\n",
    "first_sequence = merged_df_x.loc[0, \"sequence_x\"]\n",
    "merged_df_x.loc[0, \"sequence_x\"] = first_sequence\n",
    "\n",
    "# Print the final merged dataframe\n",
    "#print(merged_df_x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "353066a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_start_end_y['sequence_y'] = merged_df_x.loc[df_start_end_y.index, 'sequence_x'].values\n",
    "#print(df_start_end_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ffc2ff0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through each row in df_start_end_y and create a new DataFrame for each one\n",
    "for index, row in df_start_end_y.iterrows():\n",
    "    sequence_y = list(row['sequence_y'])\n",
    "    amino_acid_number = []\n",
    "    count = row['start_pos_y']\n",
    "    \n",
    "    for char in sequence_y:\n",
    "        if char != '-':\n",
    "            amino_acid_number.append(count)\n",
    "            count += 1\n",
    "        else:\n",
    "            amino_acid_number.append(None)\n",
    "    \n",
    "    HMM_col = pd.DataFrame({\n",
    "        'amino_acid_number': amino_acid_number,\n",
    "        'Profile_HMM': sequence_y\n",
    "    })\n",
    "\n",
    "    # Save the new DataFrame to a CSV file\n",
    "    filename = f'./Documents/Numbering_Scheme/numbering/Profile_HMM_{index}.csv'\n",
    "    HMM_col.to_csv(filename, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c1f71ed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through each row in merged_df_x and create a new DataFrame for each one\n",
    "for index, row in merged_df_x.iterrows():\n",
    "    sequence_x = list(row['sequence_x'])\n",
    "    new_count = []\n",
    "    count = row['start_pos_x']\n",
    "    \n",
    "    for char in sequence_x:\n",
    "        if char != '-':\n",
    "            new_count.append(count)\n",
    "            count += 1\n",
    "        else:\n",
    "            new_count.append(None)\n",
    "    \n",
    "    seq_col = pd.DataFrame({\n",
    "        'new_count': new_count,\n",
    "        'Profile_HMM': sequence_x\n",
    "    })\n",
    "\n",
    "    # Save the new DataFrame to a CSV file\n",
    "    filename = f'./Documents/Numbering_Scheme/numbering/seq_{index}.csv'\n",
    "    seq_col.to_csv(filename, index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "740a31ad",
   "metadata": {},
   "source": [
    "The next cell shows the core of the Numbering Scheme, here the numbering from the \n",
    "reference sequence is taken. It is aligned to the profile HMM and the query sequences,\n",
    "the umbering bounds every query sequence in such a way so that it is aligned to the Profile \n",
    "HMM but also bounded to the numbering from 1 to 263.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b9cc0149",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "# Get a list of all HMM_profile CSV files in the directory that match the pattern 'HMM_profile_*.csv'\n",
    "#HMM_file_list = glob.glob('./Documents/Numbering_Scheme/numbering/Profile_HMM_*.csv')\n",
    "HMM_file_list = sorted(glob.glob('./Documents/Profile_HMM_*.csv'), key=lambda x: int(re.findall(r'\\d+', x)[0]))\n",
    "\n",
    "# Get a list of all seq CSV files in the directory that match the pattern 'seq_*.csv'\n",
    "#seq_file_list = glob.glob('./Documents/Numbering_Scheme/numbering/seq_*.csv')\n",
    "seq_file_list = sorted(glob.glob('./Documents/seq_*.csv'), key=lambda x: int(re.findall(r'\\d+', x)[0]))\n",
    "\n",
    "# Read in the amino acid number DataFrame\n",
    "amino_acid_number = pd.read_csv('./nucleotide_number.csv')\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "# Define the renumber_index function\n",
    "def renumber_index(row):\n",
    "    global counter\n",
    "    global counternew\n",
    "    new_index = 0\n",
    "    if row['Profile_HMM'] == '-' and row['Profile_HMM_x'] == '-':\n",
    "        new_index = ''\n",
    "    elif row['Profile_HMM'] == '-' or row['Profile_HMM_x'] == '-':\n",
    "        counternew += 1\n",
    "        new_index = f\"{counter}.{counternew}\"\n",
    "    else:\n",
    "        counter += 1\n",
    "        counternew = 0\n",
    "        new_index = counter\n",
    "    return new_index\n",
    "\n",
    "# Iterate over the sorted lists of HMM_profile and seq CSV files\n",
    "for i in range(len(HMM_file_list)):\n",
    "    # Initialize the counters\n",
    "    counter = 0\n",
    "    counternew = 0\n",
    "\n",
    "    # Read in the HMM_profile file\n",
    "    Profile_HMM = pd.read_csv(HMM_file_list[i])\n",
    "    Profile_HMM = Profile_HMM.iloc[:, :3]  # Keep only the first three columns\n",
    "    # Read in the corresponding seq file\n",
    "    seq_file = pd.read_csv(seq_file_list[i])\n",
    "    seq_file = seq_file.iloc[:, :3]  # Keep only the first three columns\n",
    "    # Merge the HMM_profile and seq DataFrames on the 'Unnamed: 0' column\n",
    "    merged_df = pd.merge(Profile_HMM, seq_file, on='Unnamed: 0')\n",
    "\n",
    "    # Merge the resulting DataFrame with the 'amino_acid_number' DataFrame on the 'amino_acid_number' column\n",
    "    final_df = pd.merge(merged_df, amino_acid_number, on='amino_acid_number', how='outer')\n",
    "    #print(final_df)\n",
    "    # Sort the merged DataFrame by the 'amino_acid_number' column to maintain the original order\n",
    "    final_df = final_df.sort_values(by='amino_acid_number')\n",
    "\n",
    "    # Reset the index and drop the old index column\n",
    "    final_df = final_df.reset_index(drop=True)\n",
    "    #print(final_df)\n",
    "    # Drop unnecessary columns\n",
    "    final_clear_df = final_df.drop(columns=['Unnamed: 0'])\n",
    "    #print(final_clear_df)\n",
    "    \n",
    "    # Fill NaN values with '-'\n",
    "    # Find the minimum value in amino_acid_number that is greater than 0\n",
    "    min_amino_acid_number = final_clear_df[final_clear_df['amino_acid_number'] > 0]['amino_acid_number'].min()\n",
    "\n",
    "    # Fill the gaps for missing amino_acid_number values with '-' in the Profile_HMM_x column\n",
    "    for i in range(1, min_amino_acid_number):\n",
    "        final_clear_df.loc[final_clear_df['amino_acid_number'] == i, 'Profile_HMM_x'] = '-'\n",
    "\n",
    "    # Fill the remaining NaN values in the Profile_HMM_x column with '-'\n",
    "    final_clear_df['Profile_HMM_x'].fillna('-', inplace=True)\n",
    "\n",
    "\n",
    "    #print(final_clear_df)\n",
    "    # Convert all numeric columns to integers\n",
    "    for col in final_clear_df.columns:\n",
    "        if final_clear_df[col].dtype.kind in 'bifc':\n",
    "            final_clear_df[col] = final_clear_df[col].fillna(0).astype(np.int64)\n",
    "            #print(final_clear_df)\n",
    "        # Apply the renumber_index function to final_clear_df\n",
    "    final_clear_df['standard_numbering'] = final_clear_df.apply(renumber_index, axis=1)\n",
    "    #print(final_clear_df)\n",
    "\n",
    "\n",
    "    final_clear_df['standard_numbering'] = pd.to_numeric(final_clear_df['standard_numbering'], errors='coerce')\n",
    "    print(final_clear_df)\n",
    "\n",
    "    # Select rows with full round integer standard numbering\n",
    "    # Extract the DNA sequences from the Profile_HMM_x column\n",
    "    sequences = final_clear_df['Profile_HMM_x']\n",
    "    #print(sequences)\n",
    "    \n",
    "    # Create a new dataframe with the selected sequences\n",
    "    new_df = pd.DataFrame({'sequences': sequences})\n",
    "    #print(new_df)\n",
    "    \n",
    "    #with open(\"./sequences_new_df.txt\", \"a\") as f:\n",
    "            #print(new_df, file=f)\n",
    "        \n",
    "    # create dataframe\n",
    "    df = pd.DataFrame({'sequences': sequences})\n",
    "    #print(df)\n",
    "    # concatenate sequences into a single string\n",
    "    sequence_string = ''.join(new_df['sequences'])\n",
    "    #print(sequence_string)\n",
    "    with open(\"./Documents/sequence_string.txt\", \"a\") as f:\n",
    "            print(sequence_string, file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "20618440",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge all of the data from the curated database with the numbered sequences\n",
    "\n",
    "import pandas as pd\n",
    "nucleotide_sequence_cluser = pd.read_csv('./Documents/Numbering_Scheme/numbering/sequence_string.txt', sep=' ',header=None)\n",
    "nucleotide_sequence_cluser.columns = ['amino_acid_sequence_numbered']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "33484d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# assuming nucleotide_sequence_cluser and unique_df are defined\n",
    "nucleotide_sequence_cluster = pd.DataFrame(nucleotide_sequence_cluser)\n",
    "#parse_fa1 = pd.DataFrame(parse_fa1)\n",
    "\n",
    "# merge the two dataframes\n",
    "merged_df = nucleotide_sequence_cluster.merge(unique_df, left_index=True, right_index=True)\n",
    "\n",
    "# add a column named \"phenotype\" to the merged dataframe\n",
    "merged_df[\"phenotype\"] = \"unknown\"\n",
    "\n",
    "# save the merged dataframe to a CSV file\n",
    "merged_df.to_csv(\"merged_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9fe35c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Read the CSV file and store it in a DataFrame\n",
    "merged_df = pd.read_csv('merged_data .csv')\n",
    "merged_df = merged_df.head(2000)\n",
    "# Display the first few rows of the DataFrame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "53a83ebb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                           amino_acid_sequence_numbered name  \\\n",
      "0     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...  195   \n",
      "1     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...  163   \n",
      "2     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...    9   \n",
      "3     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMLST...   39   \n",
      "4     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...  211   \n",
      "...                                                 ...  ...   \n",
      "1995  HPETLAKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...        \n",
      "1996  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...        \n",
      "1997  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...        \n",
      "1998  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...        \n",
      "1999  HPETLVKVKDAEDQLGARVGYIELDLSSGKILESFRPEERFPLMST...        \n",
      "\n",
      "                                    amino_acid_sequence phenotype  \n",
      "0     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "1     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "2     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...       2be  \n",
      "3     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMLST...       2br  \n",
      "4     HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...       2be  \n",
      "...                                                 ...       ...  \n",
      "1995  HPETLAKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "1996  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "1997  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "1998  HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMST...   unknown  \n",
      "1999  HPETLVKVKDAEDQLGARVGYIELDLSSGKILESFRPEERFPLMST...   unknown  \n",
      "\n",
      "[2000 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Removing the TEM and leaving only the numbers\n",
    "def process_name(value):\n",
    "    if value.startswith(\"TEM-\"):\n",
    "        return value.split('-')[1]\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "# Apply the custom function to the 'name' column\n",
    "merged_df['name'] = merged_df['name'].apply(process_name)\n",
    "\n",
    "print(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3733cd2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Define a function to calculate the number of mismatches between two sequences\n",
    "def calculate_distance(seq1, seq2, max_mismatches):\n",
    "    num_mismatches = sum(1 for a, b in zip(seq1, seq2) if a != b)\n",
    "    if num_mismatches <= max_mismatches:\n",
    "        return num_mismatches\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ec88567c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to build the distance matrix\n",
    "def build_distance_matrix(merged_df, max_mismatches):\n",
    "    num_seqs = len(merged_df)\n",
    "    distance_matrix = np.zeros((num_seqs, num_seqs))\n",
    "\n",
    "    for i in range(num_seqs):\n",
    "        for j in range(i, num_seqs):\n",
    "            seq1 = merged_df.iloc[i]['amino_acid_sequence_numbered']\n",
    "            seq2 = merged_df.iloc[j]['amino_acid_sequence_numbered']\n",
    "            distance = calculate_distance(seq1, seq2, max_mismatches)\n",
    "            \n",
    "            distance_matrix[i][j] = distance\n",
    "            distance_matrix[j][i] = distance\n",
    "\n",
    "    return distance_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6600054e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "\n",
    "max_mismatches = 1\n",
    "distance_matrix = build_distance_matrix(merged_df, max_mismatches)\n",
    "\n",
    "# Define a dictionary to map phenotype values to specific colors\n",
    "phenotype_color_map = {\n",
    "    '2b': 'green',\n",
    "    '2be': 'blue',\n",
    "    '2br': 'red',\n",
    "    '2ber': 'pink',\n",
    "    'unknown': 'gray',\n",
    "}\n",
    "\n",
    "# Get the phenotype colors for each sequence\n",
    "merged_df['color'] = merged_df['phenotype'].apply(lambda x: phenotype_color_map[x])\n",
    "\n",
    "# Create graph and add edges based on distance matrix\n",
    "G = nx.Graph()\n",
    "for i, id in enumerate(merged_df['name']):\n",
    "    G.add_node(i, color=merged_df.iloc[i]['color'], label=id, font_size=5)\n",
    "\n",
    "for i in range(len(merged_df)):\n",
    "    for j in range(i+1, len(merged_df)):\n",
    "        weight = distance_matrix[i][j]\n",
    "        if weight > 0:\n",
    "            G.add_edge(i, j, weight=weight)\n",
    "            \n",
    "# Get nodes with degree > 0\n",
    "nodes_with_degree = [n for n, degree in dict(G.degree()).items() if degree > 0]\n",
    "\n",
    "# Create subgraph with nodes that have degree > 0\n",
    "subG = G.subgraph(nodes_with_degree)\n",
    "\n",
    "# Set figure size\n",
    "plt.figure(figsize=(150, 150))\n",
    "\n",
    "# Use the API below for the neworkx guidance\n",
    "# https://networkx.org/documentation/stable/reference/generated/networkx.drawing.layout.spring_layout.html\n",
    "\n",
    "# Visualize subgraph\n",
    "seed = 16\n",
    "\n",
    "pos = nx.spring_layout(subG, k=0.15, iterations=100, scale=1, seed=seed)\n",
    "nx.draw_networkx_nodes(subG, pos, node_color=[subG.nodes[n]['color'] for n in subG.nodes], node_size=500)\n",
    "nx.draw_networkx_edges(subG, pos, width=1, alpha=1)\n",
    "smaller_labels = {i: merged_df.iloc[i]['name'] for i in nodes_with_degree}\n",
    "nx.draw_networkx_labels(subG, pos, smaller_labels, font_size=5)\n",
    "\n",
    "output_filename = \"2000.png\"\n",
    "plt.savefig(output_filename, format=\"png\", dpi=150)\n",
    "\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1b6b9627",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-22-6742a441051e>:36: RuntimeWarning: divide by zero encountered in power\n",
      "  fitted_power_law = lambda x: np.exp(popt[0]) * np.power(x, popt[1])\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAEaCAYAAADtxAsqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3de5xN5f7A8c/XUDNF6YLjUoZIctcISRGiEiqdKEJKdJSuJ53uR4pDKUWSim6kC7oeKSmp38mIyiX3KbcQSQq5fH9/PGvGttu3mdl79mW+79drv2b2ujzr2Xutvb97redZ30dUFWOMMSYSJeJdAWOMMcnDgoYxxpiIWdAwxhgTMQsaxhhjImZBwxhjTMQsaBhjjImYBY1CEJGJIvJQnLYtIvKCiPwiIl/FeFutRGR9LLdRVETkEhFZJyK7RKRREW73KhH5sKi257PdFiKy0nu9XSJYPlNEVERKFmBbJ3vbSStYbeP7mUoVIpIjIm29//8lIhOiWX5KBQ3vzdosIkf7TLtWRObEsVqxcjbQDqiiqmfGuzJJZCQwUFVLq+rCWGwg0Bevqr6iqufHYnth/Bt4ynu90/1n+n7BFJaq/uht50A0yituChOwg1HVh1X12miWn1JBw1MSGBTvSuRXAX6dVQVyVPX3WNQnhVUFlsS7EkUoIV6vd2Yct++baH4Rx1MivI5UDBojgNtFpKz/jECRVkTmiEhuJO4tIvNEZJSI7BCRNSJyljd9nYhsEZFefsWeKCKzROQ3EflURKr6lH2aN2+7iCwXkb/7zJsoIk+LyPsi8jvQOkB9K4nI2976q0TkOm96X2AC0Ny7HPBggHV7i8jnIjLSu4S1VkQuCFe2Ny/Dq98vIrIUaBKgXm+KyFav3Jt85p0pItkistM763ss0E4SkeNE5F2vjF+8/6v41X+N976uFZGrgpRzpoh86e2vTSLylIgcEWC5I0VkF5AGfCMiq73pKiI1/PbLQ97/rURkvYjc5u37TSLSx+99elREfhCRX733OwP4zFtkh7d/mufuD591zxKR+d5680XkLJ95c0RkiHcs/iYiH4rIiYFev7f8dd4+3O7t00re9NVAdeAdrx5H+q33EnCyz/x/+sy+SkR+FJGfReRun3VKiMhgEVktIttEZKqIHO/NO+zz5b2OoSIyD/jDq4t/3RuJyNfe63wNSPeb31FEFnn79wsRqe8zr7GILPTWfV1EXguw7+4UkZ+AFyIoL+hxHaDewfY9ItLMK3uHiHwjIq181gu1b4MdN7nfSduBB0TkFBGZ7b3/P4vIKxLg+87b3gMi8nKQ8s/1jpl6PsuXF5HdIlIu2GtHVVPmAeQAbYG3gIe8adcCc7z/MwEFSvqsMwe41vu/N7Af6IP7cnkI+BEYAxwJnA/8BpT2lp/oPT/Hm/8E8Lk372hgnVdWSaAx8DNQx2fdX4EWuOCdHuD1fAqMxX2QGgJbgTY+df08xHvRG9gHXOe9lgHARkAiKHsYMBc4HjgJWAys9+aVABYA9wFH4L4I1gDtvflfAj29/0sDzYLU7wTgMuAooAzwOjDd573bCdTynlfMfd8ClHMG0Mx7jzOBZcDNId4XBWqEeD7R59hp5R0P/wZKARfivvyO8+aPwR0/lb33+CzvOMjkr8dZ3v7y3tdfgJ5evbt7z0/wOSZXA6cCGd7zYUFez3m446qxt+0ngc/8PxPhPjM+z3Pr/qy37QbAXqC2N/9m4P+AKt72ngEmB/p8efX+Eajjvc5Sfts+AvgBuMV7f7vijtnc978xsAVo6r2/vbz6Humz7iBv3UuBPwPsu+He8hlhygt5XAd434Lt+8rANtyxUgJ3CXkbUC7cvvV///y+k2703sMMoIZX7pFAOVwweDzQPgUeAF4OUf5YYLjP80HAOyG/Z+Px5R6rB4eCRl3cF3I58h80VvrMq+ctX8Fn2jagoc8XzBSfeaWBA7gv2iuAuX71ewa432fdF0O8lpO8ssr4THsEmOj/JRRk/d7AKp/nR3mv5W8RlL0G6OAzrx+HgkZT4Ee/bd0FvOD9/xnwIHBiPvddQ+AX7/+jgR24oJKRz3JuBqaFmJ/foLHb73jZggtSJbx5DQJsI9Bxlre/cMHiK791vgR6+xyT9/jMuwH4b5DX8xzwH79jcB+Q6fuZCPeZCVD3Kj7TvgK6ef8vw/tx4T2v6G2vJIGDxr9DbPscfH7IeNO+8Hn/nwaG+K2zHDjXW3eD37qf++27P/H5MRamvJDHtd/0UPv+TuAlv2kzgV7h9m2I4+ZH/+34ld8FWBhonxI+aDTF/bgt4T3PBv4eanupeHkKVV0MvAsMLsDqm33+3+2V5z+ttM/zdT7b3QVsByrhriU39U5Rd4jIDuAq3Jf2X9YNoBKwXVV/85n2A+6XTKR+8qnbH96/pSMou5Jf3X7w+b8qUMnvdf0LqODN74v7FfW9d9mlY6CKichRIvKMd3q/ExdsyopImrp2miuA/sAmEXlPRE4LUs6p4i5t/eSV8zAQ9FJOAWxT1f0+z//AvYcn4s7SVhegzEoc/p7CX/ftTz7/524zbFneMbiN/B0ngQTbflVgms++X4b7AVKBwMId4xvU+7by+B9rt/kdayd56wVa139bW1V1T4TlhTuufYXa91WBy/3KORsXXHNFum8Dvi7vEtIUEdngHfMvU8BjXlX/B/wOnOt9xmoAb4daJyWDhud+3KUZ3w9PbqPxUT7TfL/EC+Kk3H9EpDTu0sNG3I7+VFXL+jxKq+oAn3WV4DYCx4tIGZ9pJ+N+XRVWuLI34fO6vHm51gFr/V5XGVW9EEBVV6pqd6A87tLAG+LTm83HbUAtoKmqHoP75QggXjkzVbUd7sP2Pe5ySSBPe/NreuX8K7eMCP1BwY6Hn4E9wCkB5oXar+De/6p+0wq6bw8ry3uvT8hHWeHq6m8dcIHf/k9X1WDbC1X+JqCyiPjuL/9jbajfto5S1clB1vU9ZgNtO1R5IY9rP6H2/TrcmYZvOUer6rAQ70Ow+gab/og3rb53zPcgsmM+WPmTvDJ6Am/4Bdq/SNmgoaqrgNeAm3ymbcV9mHqISJqIXEPgHZ8fF4rI2eIaX4cA/1PVdbgznVNFpKeIlPIeTUSkdoT1X4c7VX9ERNK9Bru+wCuFrG8kZU8F7hLXWF0Fdz0111fATq+BMcN7H+uKSBMAEekhIuVU9SDuEhO4X6L+yuDO2naIa0i9P3eGiFQQkU7eF+BeYFeQMnLL2Qns8n4pDQiyXDCLgCu919EBd6kiLO/1PQ885jWgpnkNl0fi2ocOEqDh1/M+7ti4UkRKisgVwOm4Yya/XgX6iEhDb9sP447BnAjX3xyinoGMA4aK1+FDRMqJSOf8VNjHl7jr9Td578OlgG/38WeB/iLSVJyjReQi78fOl7hjYqC3bme/dQMJVV7I49pXmH3/MnCxiLT3pqeLa5Sv4l9OAOGOm1xlcJ+JHSJSGbgjgrJDlf8ScAkucLwYrpCUDRqef+Ouj/u6Dvcmb8M10H1RyG28ivvC245rlL0KwLv0cz7QDfdr8CcONcpFqjvuOuRGYBquPWRWIesbSdkP4i4TrAU+xB1UAKjrg38xrg1iLe5X1wTgWG+RDsAScT2VnsBdCw/0y+VxXKPez7iG1f/6zCuBOxPZiHtfz8Vd+w3kduBKXIeEZ3E/FPJjkPd6ci8f/uVehhBuB74D5nv1HI67NvwHMBSY512iaOa7kqpuAzriXuM24J9AR1X9OZ91R1U/Bu4F3sT9+j4Fd8xF6hHgHq+et0ew/BO4yxcfishvuH3XNH+1dlT1T1wDdm9cR4ArcJ1Ycudn4z6vT3nzV3nL+q7bF7fveuCC7t4Q2wtVXrjj2l+wfb8O6Iw7492KO/O4gwi+a8MdNz4exDXq/wq8h897VpDyVXU98DXuTGRuuHJye9IYY0xSE5H/AeNU9YV41yXZiMjzwEZVvSfcsnG/UcQYYwpCRM7F9X76GXeWWJ/Dz1hNBEQkE3fWFlFanVS/PGWMSV21gG9wl2luA7qq6qb4Vim5iMgQ3H1YI1R1bUTr2OUpY4wxkbIzDWOMMRGzoGGMMSZiKdkQfuKJJ2pmZma8q2GMMUljwYIFP6tq8ESFnpQMGpmZmWRnZ8e7GsYYkzRExD+1TUB2ecoYY0zELGgYY4yJmAUNY4wxEUv4Ng0vwd8gXOrfj1X16ThXyUTJvn37WL9+PXv2hEyqaYyJovT0dKpUqUKpUqUKtH5cgoaX56QjsEVV6/pM74BLiJYGTFDVYaq6DJeZsgTB02ObJLR+/XrKlClDZmYmh2e4NsbEgqqybds21q9fT7Vq1QpURrwuT03EZUPNIyJpuCEUL8Clie4uIqd78zrhRuX6uGiraWJpz549nHDCCRYwjCkiIsIJJ5xQqLP7uAQNVf0Ml07Y15m44UnXeGmPp+BSDKOqb6vqWXhpx03qsIBhTNEq7GcukRrCK3P4sIbrcSNztRKR0SLyDG7wmoBEpJ+IZItI9tatW2NdV5Mi0tLSaNiwYd4jJyeH7OxsbrrJjd01Z84cvvji0JAr06dPZ+nSpfneTunS4Ub0jK9WrVol1b1Ns2fPpnHjxtStW5devXqxf//+w+bPnz+ftLQ03njjjZis72/Xrl1kZWVRvXp1Nm7c+Jf5Xbt2Zc2aNX+Zvm3bNlq3bk3p0qUZOHBg0PK3b99Ou3btqFmzJu3ateOXX34B4M8//6RPnz7Uq1ePBg0aMGfOnLx12rZtm7dcNCVS0AgU/lRV56jqTap6vaqOCbayqo5X1SxVzSpXLuxNjcYAkJGRwaJFi/IemZmZZGVlMXr0aCB6QSOR+H9BJpuDBw/Sq1cvpkyZwuLFi6latSqTJk3Km3/gwAHuvPNO2rdvH5P1/e3fv5+///3v9OzZkxEjRtC5c2d27tyZN3/JkiUcOHCA6tX/OiBfeno6Q4YMYeTIkSG3MWzYMNq0acPKlStp06YNw4a50WOffdY183733XfMmjWL2267jYMHDwLQs2dPxo4dG9FryI9EChrrOXyM3yq4kdsiJiIXi8j4X3/9NaoVM8XLnDlz6NixIzk5OYwbN45Ro0bRsGFDPv30U95++23uuOMOGjZsyOrVq1m9ejUdOnTgjDPOoGXLlnz//fcArF27lubNm9OkSRPuvffegNvJycnhtNNOo1evXtSvX5+uXbvyxx9/APDxxx/TqFEj6tWrxzXXXMPevXv56quvuPTSSwGYMWMGGRkZ/Pnnn+zZsyfvCylYfXr37s2tt95K69atufPOO0O+/gEDBpCVlUWdOnW4/343Cm8k2w6mZcuWLFq0KO95ixYt+Pbbb0OuE8q2bds48sgjOfXUUwFo164db775Zt78J598kssuu4zy5ctHbf358+dTv3599uzZw++//06dOnVYvHgxANdffz0XXHABgwYN4rLLLuPuu++mW7du7Nu3D4BXXnmFzp0Dj4h79NFHc/bZZ5Oenh7yNc+YMYNevXoB0KtXL6ZPdwNMLl26lDZt2gBQvnx5ypYtm3fG2KlTJyZPnhyy3IJIpC6384GaIlINN453N9wwnhFT1XeAd7Kysq6LQf1MLN18M/h8sURFw4bw+OMhF9m9ezcNGzYEoFq1akybNi1vXmZmJv3796d06dLcfrsbCbVTp0507NiRrl27AtCmTRvGjRtHzZo1+d///scNN9zA7NmzGTRoEAMGDODqq69mzJigJ8gsX76c5557jhYtWnDNNdcwduxYBg4cSO/evfn444859dRTufrqq3n66acZOHAgCxcuBGDu3LnUrVuX+fPns3//fpo2dSOu9uvXL2B9AFasWMFHH31EWlpayPdk6NChHH/88Rw4cIA2bdrw7bff0rhx47DbHjduHAD9+/c/rLxrr72WiRMn8vjjj7NixQr27t1L/fr1//I+XHHFFQHrM2fOHMqWLZv3/MQTT2Tfvn1kZ2eTlZXFG2+8wbp17sr2hg0bmDZtGrNnz2b+/PkByyvI+k2aNKFTp07cc8897N69mx49elC3ruv4+dxzzx1WfpcuXejSpUve83nz5tG9e/eAdYnU5s2bqVixIgAVK1Zky5YtADRo0IAZM2bQrVs31q1bx4IFC1i3bh1nnnkmxx13HHv37mXbtm2ccMIJhdq+r3h1uZ0MtAJOFJH1uPGpnxORgcBMXJfb51V1ST7LvRi4uEaNGtGusklRuZenCmLXrl188cUXXH755XnT9u51Q1TPmzcv79drz549g/66P+mkk2jRogUAPXr0YPTo0bRr145q1arl/RLu1asXY8aM4eabb6ZGjRosW7aMr776iltvvZXPPvuMAwcO0LJly5D1Abj88svDBgyAqVOnMn78ePbv38+mTZtYunQp9evXD7lt+Guw8N3ukCFDGDFiBM8//zy9e/f+yzK1atWKeD+ICFOmTOGWW25h7969nH/++ZQs6b7Kbr75ZoYPHx7ydRZ0/fvuu48mTZqQnp6ed/kyEps2bSJWl8yvueYali1bRlZWFlWrVuWss87Key3gzj42btyY/EFDVQOGXVV9nxCN3RGUa2caySrMGUEiOnjwIGXLlg36ZRdJLxX/ZUSEUAOjtWzZkg8++IBSpUrRtm1bevfuzYEDBxg5cmTY+hx99NF5/7dv357NmzeTlZXFhAkT8qavXbuWkSNHMn/+fI477jh69+6d1z0z1LZDOeqoo2jXrh0zZsxg6tSpARvc83OmAdC8eXPmzp0LwIcffsiKFSsAyM7Oplu3bgD8/PPPvP/++5QsWfKwX/4FXX/79u3s2rWLffv2sWfPnsPez1AyMjLy3sNp06bx4IMPAjBhwgSysrIiKqNChQps2rSJihUrsmnTprxLZyVLlmTUqFF5y5111lnUrFkz7/mePXvIyMiIaBuRSqQ2DWMSTpkyZfjtt98CPj/mmGOoVq0ar7/+OuBunPrmm28Ad91+ypQpgLumHcyPP/7Il19+CcDkyZM5++yzOe2008jJyWHVqlUAvPTSS5x77rkAnHPOOTz++OM0b96ccuXKsW3bNr7//nvq1KkTsj7+Zs6cyaJFiw4LGAA7d+7k6KOP5thjj2Xz5s188MEHefNCbTuca6+9lptuuokmTZpw/PHH/2V+7plGoId/wADyLs/s3buX4cOH553lrF27lpycHHJycujatStjx479S8Ao6Pr9+vVjyJAhXHXVVWHbhXzVrl07b19ecsklea8r0oAB7rJobmP9pEmT8tpI/vjjD37//XcAZs2aRcmSJTn99NMBt/9/+uknoj1MREoFDWsIN9F28cUXM23aNBo2bMjcuXPp1q0bI0aMoFGjRqxevZpXXnmF5557jgYNGlCnTh1mzJgBwBNPPMGYMWNo0qQJoY7H2rVrM2nSJOrXr8/27dsZMGAA6enpvPDCC1x++eXUq1ePEiVK5H2pNW3alM2bN3POOecAUL9+ferXr593xhKsPpFq0KABjRo1ok6dOlxzzTV5l84i2fa4cePy2jX8nXHGGRxzzDH06dMnX/UJZsSIEdSuXZv69etz8cUXc95554Vd58ILL8zrDpvf9V988UVKlizJlVdeyeDBg5k/f35eW1E4F1100WFdYf1lZmZy6623MnHiRKpUqZLXO+/aa6/NOysbPHgws2bNombNmsyaNYvBgwcDLvg1btyY2rVrM3z4cF566aW8chcsWECzZs0Ou1wVDSk5RnhWVpYmU5/z4mrZsmXUrl073tWIm5ycHDp27JjXCyeVbdy4kVatWvH9999TokRK/VYNa/fu3bRu3Zp58+ZF1KYULYMGDaJTp055vat8BfrsicgCVQ17+lO89p4xpsi9+OKLNG3alKFDhxa7gAGuTePBBx9kw4YNRbrdunXrBgwYhWVnGiZuivuZhjHxUpgzjUS6T6PQCtPldvrCDYyYuZyNO3ZTqWwGd7SvRZdGlaNfSWOMSWIpda6oqu+oar9jjz02X+tNX7iBu976jg07dqPAhh27ueut75i+sGhPJ40xJtGlVNAoqBEzl7N734HDpu3ed4ARM5fHqUbGGJOYLGgAG3fsztd0Y4wprlIqaBT0Po1KZQPfMRlsukkduanR69SpQ4MGDXjsscfysoQGk5OTw6uvvlpENTQmsaRU0Chom8Yd7WuRUerw/tMZpdK4o32taFbPJKDc3FNLlixh1qxZvP/++3lpHoKxoGGKs5QKGgXVpVFlHrm0HpXLZiBA5bIZPHJpPes9lWCmL9xAi2GzqTb4PVoMmx31jgrly5dn/PjxPPXUU6gqOTk5tGzZksaNG9O4ceO8cTUGDx7M3LlzadiwIaNGjQq6nDGpKKW63BZGl0aVLUgksNwebrkdFnJ7uAFR3W/Vq1fn4MGDbNmyhfLlyzNr1izS09NZuXIl3bt3Jzs7m2HDhjFy5EjeffddwOX/CbScManIgoZJCqF6uEU72Ofe8Lpv3z4GDhzIokWLSEtLy8uE6i/S5YxJBSkVNGw8jdRVVD3c1qxZQ1paGuXLl+fBBx+kQoUKfPPNNxw8eDDo6GqjRo2KaDljUkFKtWkUtCHcJL6i6OG2detW+vfvz8CBAxERfv31VypWrEiJEiV46aWXOHDAnen4p0sPtpwxqSilgoZJXbHq4ZY73GudOnVo27Yt559/ft642DfccAOTJk2iWbNmrFixIm/Qnfr161OyZEkaNGjAqFGjgi5nTCqyhIUmbvKbsNDygxkTHZaw0BQL1sPNmPizy1PGGGMillJBw4Z7NcaY2EqpoGG9p5JPKrapGZPICvuZS6mgYZJLeno627Zts8BhTBFRVbZt21aoe4msIdzETZUqVVi/fj1bt26Nd1WMKTbS09OpUqVKgde3oJFgilO30lKlSlGtWrV4V8MYkw8WNBJIUSXlM8aYgrI2jQRiw84aYxKdBY0EYsPOGmMSnQWNBGLDzhpjEl1KBY1kv7nPhp01xiS6lAoayX5znw07a4xJdNZ7KsFYUj5jTCJLqTMNY4wxsWVBwxhjTMQsaBhjjImYBQ1jjDERs6BhjDEmYtZ7KsUVpwSIxpjYs6CRwiwBojEm2uzyVAqzBIjGmGizoJHCLAGiMSbaEv7ylIh0AS4CygNjVPXDOFcpaVQqm8GGAAEikgSI1hZijAkkLmcaIvK8iGwRkcV+0zuIyHIRWSUigwFUdbqqXgf0Bq6IQ3WTVkETIOa2hWzYsRvlUFvI9IUbYlhbY0wyiNflqYlAB98JIpIGjAEuAE4HuovI6T6L3OPNNxEqaAJEawsxxgQTl8tTqvqZiGT6TT4TWKWqawBEZArQWUSWAcOAD1T162Blikg/oB/AySefXLCKPfAAVKoEvXrBkUcWrIwEU5AEiNYWYowJJpEawisD63yer/em3Qi0BbqKSP9gK6vqeFXNUtWscuXK5X/rBw7ARx/B9dfDKafA6NHwxx/5LycF2GBQxphgEiloSIBpqqqjVfUMVe2vquNCFlCYQZjS0mDuXPjwQxc0Bg2CzEwYPhx27sx/eUnMBoMyxgSTSEFjPXCSz/MqwMb8FFDoQZhEoF07+PRT+OwzaNwYBg92weOBB2D79oKVm2RsMChjTDCiqvHZsGvTeFdV63rPSwIrgDbABmA+cKWqLslv2VlZWZqdnR2dis6fDw8/DNOnQ+nS8I9/wC23QIUK0SnfGGMSgIgsUNWscMvFq8vtZOBLoJaIrBeRvqq6HxgIzASWAVPzGzBiMkZ4kyYwbRp8+y107AgjRrgzj0GDYP366G3HGGOSQNzONGIpqmca/lasgGHD4KWX3OWs3r3dJazq1WOzPWOMKQIJfaaR1E49FZ5/HlatguuugxdfdNOuvhqWLYt37YwxJqbCBg0ROUVEjvT+byUiN4lI2dhXLf9icnkqmKpVYcwYWLsWbr4Z3nwT6tSByy+HRYtiv31jjImDSM403gQOiEgN4DmgGvBqTGtVQIXuPVUQFSvCyJHwww/wr3+5LruNGrn2jy+/LLp6GGNMEYgkaBz0GqkvAR5X1VuAirGtVhI68UR46CEXPB56CP7v/+Css6BNG/jkE0jBtiNjTPETSdDYJyLdgV7Au960UrGrUsEV6eWpYMqWhbvvhpwcePRRWLoUzjsPzj4b3n8/JYPH9IUbaDFsNtUGv0eLYbMtsaExKSySoNEHaA4MVdW1IlINeDm21SqYuFyeCqZ0abj1VtfmMXas65570UVwxhnw1ltw8GC8axgVlhHXmOIlbNBQ1aXAncDX3vO1qjos1hVLGenpMGCA6231/POwaxdcdhnUqwevvAL798e7hoViGXGNKV4i6T11MbAI+K/3vKGIvB3riqWcUqWgTx/XLXfyZChRAnr0gNNOgwkT4M8/413DArGMuMYUL5FcnnoAl7Z8B4CqLsL1oEo4CdGmEU5aGnTrBt9841KTHHecu9/jlFPgySdhd3J92VpGXGOKl0iCxn5V9f8WTsjW3IRq0winRAno3Bm++gpmzoRq1eCmm9zf//wHfvst3jWMiGXENaZ4iSRoLBaRK4E0EakpIk8CX8S4XsWHCJx/vsuq++mn0KAB3Hmnu3nw3/+GX36Jdw1Dsoy4xhQvYXNPichRwN3A+bgxL2YCQ1R1T+yrVzAxzT1VFL76ymXWnTEDypQ5lFm3fPl418wYk6IizT1lCQsT2bffuuAxdarrhdWvH9x+O1SpEu+aGWNSTKGDhoi8Q4i2C1XtVPDqxYbX0+viGjVqXLdy5cp4Vyd6li93mXVfftm1hfTu7S5hWWZdY0yURCPL7UjgUWAtsBt41nvsAhZHo5LRllQN4flRqxa88AKsXAl9+8LEiZZZ1xgTF0GDhqp+qqqfAo1U9QrvC/kdVb0SOLvoqmjyZGa6u8vXrnWDQOVm1v373y2zrjGmSETSe6qciORdB/HSiJSLXZVMWJUqubxWOTlw112uy26jRnDxxS5RojHGxEgkQeMWYI6IzBGROcAnwKCY1spEplw5GDrUZdYdMsSlYm/eHNq2hTlzUjI5ojEmviLJPfVfoCYuUAwCaqnqh7GumMmHsmXhnnvcmcfIkbBkCbRundKZdY0x8RFJ7qlSwPXAvd7jOm9awkmKNCKxVLo03Haba/MYMyZlM+saY+InkstTT7VQXzwAABt9SURBVANnAGO9xxnetISTsr2n8is9HW64wfW2eu45l5IkhTLrGmPiJ5Kg0URVe6nqbO/RB2gS64qZKDjiCLjmGvj+e3j11ZTJrGuMiZ9IgsYBETkl94nXk+pAiOVNoklLg+7dXWbdadNcG8h110GNGkmZWdcYEz+RBI07gE+83lOfArOB22JbLRMTJUpAly4wfz78978uKWISZtY1xsRPRLmnRORIoBYuYeH3qro31hUrjJTJPVUUPvsMHnoIZs1yY3vcfDPceKP7vwhNX7iBETOXs3HHbiqVzeCO9rUsU64xRSgaaUR8nQHUBRoAV4jI1YWpnEkg55wDH34I//sftGwJ99/vzkDuugu2bCmSKtg448Ykj0i63L6Ey0N1Nq4BvAkQNhqZJHPmmS4V+zffwIUXwvDhLm3JzTe7rrsxZOOMG5M8IjnTyAJaqOoNqnqj97gp1hUzcVK/PkyZ4hIh/v3v8NRTbija66+HNWtiskkbZ9yY5BHRyH3A32JdkWgo9jf3RVOtWi6b7sqVrttuDDPr2jjjxiSPSILGicBSEZkpIm/nPmJdsYKwm/tioFo1ePppd5Zx003wxhtRz6xr44wbkzwiGe713EDTvbTpCcl6T8XQ1q3w+OPu/o7ffoOOHeHuu6FZs0IVa72njIkvG+7VgkZs/fKLa+94/HHYvh3atHFJE889F0TiXTtjTD5Fu8utMYc77ji4916Xln3ECFi82GXWbdkSPvjAMusak6IsaJjCKV0abr/dZdZ96in48UfXZTcryzLrGpOCggYNEfnY+zu86KpjklZGBvzjH7Bqlcusu3OnZdY1JgWFOtOo6DWCdxKRRiLS2PdRVBU0SSY3s+6yZS5YiFhmXWNSSNCGcBHpCvTF3Qnu36qsqnpejOtWYNYQnkAOHnR3mg8dCgsWwEknwR13wLXXurMTY0xCKHRDuKq+oaoXAP9R1dZ+j4QNGCbBlCgBl1ziMut+8AGcfPKhzLojRlhmXWOSTCRjhA8RkU4iMtJ7dCyKipkUIwIdOsDcuTBnjmvr+Oc/XX6rIUNgx45419AYE4FIEhY+AgwClnqPQd40Y/JPxN3LMWsW/N//wVlnwX33uTOQf/3L3TxojElYkXS5vQhop6rPq+rzQAdvWpEQkeoi8pyIvFFU2zRFpGlTeOcdl47kggtg2DCXlv2WW2CDpUU3JhFFep9GWZ//C53YSUSeF5EtIrLYb3oHEVkuIqtEZDCAqq5R1b6F3aZJYA0awGuvwdKlcPnlLkVJ9eowYADk5MS7dsYYH5EEjUeAhSIyUUQmAQuAhwu53Ym4M5Y8IpIGjAEuAE4HuovI6YXcjkkmp50Gkya5zLp9+sDzz7txzHv3huU2toYxiSCShvDJQDPgLe/RXFWnFGajqvoZsN1v8pnAKu/M4k9gCtC5MNsxiWP6wg20GDabaoPfo8Ww2aFH5atWDcaNc5l1b7wRpk6F2rXhiivcIFHGmLiJ6PKUqm5S1bdVdYaq/hSjulQG1vk8Xw9UFpETRGQc0EhE7gq2soj0E5FsEcneao2pCaXAw7lWrgyjRrlLVHfe6brsNmwInTq54WmNMUUukXJPBUqNqqq6TVX7q+opqhq015aqjlfVLFXNKleuXAyrafKr0MO5li8PjzzikiP++98wb55Lxd6uHXz6qSVHNKYIJVLQWA+c5PO8CrAxPwXYyH2JKWrDueZm1s3Jgf/8B777Dlq1cpl1//tfCx7GFIGQQUNESvj3cIqh+UBNEakmIkcA3YB8jRBoI/clpqgP51qmjEtFsnat62n144+uy26TJjBtmmXWNSaGQgYNVT0IfCMiJ0dzoyIyGfgSqCUi60Wkr6ruBwYCM4FlwFRVXZLPcu1MIwHFbDjXjAwYONBl1p0wwd1VfumlUL8+vPqqZdY1JgYiGe51NtAE+Ar4PXe6qnaKbdUKzhIWJp4iGc51/37X02roUHfPR40aMHgw9Ozpsu8aY4KK2nCvNka4STq5mXUfegi+/tpl1r3zTpey3TLrGhNQ1IZ79YJDDlDK+38+8HWhaxgDdnnKAIcy62Znw/vvu6AxcKC7/2PkSNi1K941NCZpRZKw8DrgDeAZb1JlYHosK1VQ1hBuDiPiGsg//xw++QTq1nUN6FWrWmZdYwooki63/wBaADsBVHUlUD6WlTImqkRc19yPPoIvvzyUWbdqVcusa0w+RRI09nppPQAQkZKAdYg3yalZM5dZd+FCaN/eZdbNzIRbb4WN+botyJhiKZKg8amI/AvIEJF2wOvAO7GtVsFYm4aJWMOGrqfV0qXQtSuMHu3aPCyzrjEhRdJ7qgRurPDzcak+ZgITNNyKcWS9p0y+rVnj7jJ/4QU4cAB69IC77oJahbyXxJgkEc3eUweBScAQ4EFgUiIHDGMKpHr14Jl1v/023rUzJmFE0nvqImA1MBp4ClglIhfEumLGxEWgzLoNGrjMul99Fe/aGRN3kbRpPAq0VtVWqnou0BoYFdtqFYy1aZio8c2s++CDrttu06aWWdcUe5EEjS2qusrn+RpgS4zqUyh2n4aJuuOOc91zf/gBhg93l6patYJzzrHMuqZYCho0RORSEbkUWCIi74tIbxHphes5Nb/IamhMIihTBv75T3fZavRo9zc3s+706ZZZ1xQboc40LvYe6cBm4FygFbAVOC7mNTMmEWVkuIby1avh2WfdXeWXXOLaPSZPdj2vjElhYbvcJiPrcmuKzP798NprLrPusmVQs6bLrNujh2XWNUklal1uvUGRHhORt0Tk7dxHdKoZXdYQbopcyZJw1VWweDG8+SaULg19+7rgMXYs7NkT7xoaE1WR3Nz3DfAc8B2Qd+HWUqMbE4Cq66Y7dCh88QX87W9w++1w/fUuoBiToKJ2pgHsUdXRqvqJqn6a+4hCHY1JPSJw4YWHMuvWqeOCRmamG9/DMuuaJBdJ0HhCRO4XkeYi0jj3EfOaGZPM/DPrNm8O997rMuvefbdl1jVJK5KgUQ+4DhiGu9HvUWBkLCtlTErJzaz79ddw/vnupkHLrGuSVCRB4xKguqqeq6qtvcd5sa6YMSmnUSN4/XVYsgQuu+xQZt0bbrDMuiZpRBI0vgHKxroi0WC9p0xSqF0bXnwRVqyA3r1hwgTX26pPHzfNmAQWSdCoAHwvIjMTvcutpRExSaV6dXjmGZdZ9x//cPd7nHYadOtmmXVNwoqky+25gaYncg8q63JrktKWLS7D7lNPwa5d0LmzazRv0iTeNTPFQKRdbu2OcGMSzfbt8OST8MQT8MsvrvH87rtdkkRjYiSad4T/JiI7vcceETkgIjujU01jzF8cfzzcf/+hzLqLFsG557qgMXOmZdY1cRXJyH1lVPUY75EOXIYbjMkYE0u5mXXXrnU9rdauhQ4d4MwzYcYMy6xr4iKShvDDqOp0wLrcmiIxfeEGWgybTbXB79Fi2GymL9yQUOUViaOOOjyz7vbt0KWLy6w7ZYpl1jVFKpLLU5f6PLqKyDDAzo9NzE1fuIG73vqODTt2o8CGHbu5663vCvxFH+3yitwRR8C118Ly5fDyyy5YdO/uuvC+8ALs2xfvGppiIJIzjYt9Hu2B34DOsayUMQAjZi5n977Df0Xv3neAETOXJ0R5ceObWfeNN1wixGuucfd6PP20ZdY1MRVJm0Yfn8d1qjpUVRNyuFe7uS+1bNyxO1/Ti7q8uCtRwt1ZvmABvPceVKrk7i6vXh0efdR12zUmykIN93pfiMe9RVnJSNnNfamlUtmMfE0v6vISRm5m3XnzYPZsd7kqN7Pu0KGWWddEVagzjd8DPAD6AnfGuF7GcEf7WmSUSjtsWkapNO5oXyshyks4ItC6NXz8sRvLo1kzuOcel1n3nnvg55/jXUOTAoIGDVV9NPcBjAcygD7AFKB6EdXPFGNdGlXmkUvrUblsBgJULpvBI5fWo0ujyglRXkJr3hzeffdQZt2HH3bB47bbYNOmeNfOJLGQd4SLyPHArcBVwCTgCVX9pYjqVmB2R7gxfpYtcynZX33VNaT37evuAalaNd41Mwmi0HeEi8gIYD6ut1Q9VX0gGQKGMSaA3My6y5fD1Ve7+z1q1HC9riyzrsmHUG0atwGVgHuAjT6pRH6zNCLGJKlTToHx411m3RtugMmTXUDp3h2++y7etTNJIFSbRglVzfBLI3JM7vOirKQxJsqqVHEJEXNy4I47XPtH/fruTvP58+NdO5PA8p1GxBiTQipUgGHDXHLEBx6Azz5zua3at4e5c+NdO5OALGgYYwJn1j3nHJdd98MPLbOuyWNBwxhziG9m3SeecG0f7dtD06bw9tuWWddY0DDGBHDUUXDTTbBqlWs437bNjSTYsKFl1i3mEn7kPhE5GhgL/AnMUdVXwq1j92mYXNMXbmDEzOVs3LGbSmUzuKN9rcNu5vOf3/q0cnzy/dagz8Ot7z8/Zezf74LFww+7ez5OPRXuusslTixVKt61i7t4HwfR2H5CD/cqIs8DHYEtqlrXZ3oH4AkgDZigqsNEpCewQ1XfEZHXVPWKcOVb0DBwKBW6b2bbjFJpeXeBB5ofTrj1feenpIMHYdo0eOgh1+5RtSrceSf06QPp6fGuXVzE+ziI1vajNtxrjEwEOvhOEJE0YAxwAXA60F1ETgeqAOu8xeyc2EQsXCr0QPPDCbd+UqZaz4/czLpff+266VaseCiz7mOPwe+/hy8jxcT7OCjq7cclaKjqZ8B2v8lnAqtUdY2q/onLcdUZWI8LHBD6DvZ+IpItItlbt26NRbVNkgmXCr2wKdZTLtV6fojARRe5xIgff+xuELztNpdZ9+GHoRgNTxDv46Cot59IDeGVOXRGAS5YVAbeAi4TkaeBd4KtrKrjVTVLVbPKlSsX25qapBAuFXphU6ynbKr1/BCB8847lFm3aVO4+2532eree4tFZt14HwdFvf1EChoSYJqq6u/eAFADwjWC2yBMxle4VOiB5ocTbv2USrWeX76Zddu2dWN5ZGa6sT1SOLNuvI+Dot5+IgWN9cBJPs+rABvzU4ANwmR8hUuFHmh+j2Ynh3webv2UbgSPVKNGbhjaxYvhkktg1CioVg3+8Q9382CKifdxUNTbj1uXWxHJBN7N7T0lIiWBFUAbYAMuw+6Vqrokv2Vb7yljEsjq1e4u84kT3Z3lPXu67ro1a8a7ZsZHQveeEpHJwJdALRFZLyJ9VXU/MBCYCSwDpuY3YNjlKWMSUG5m3dWrD2XWPe00uPJKdzZikkrC39xXEHamYUwC27zZXbIaMwZ27XKZde++G7LC/sg1MZTQZxrGmGLMN7Pu/ffDnDnQpAl06ACffx7v2pkwUipo2OUpY5LI8ce7dOw//OCCyNdfQ8uWLrPurFmWWTdBpVTQsN5TxiShY45xqUhyclxm3dWr4fzzD2XWteCRUFIqaBhjklhuZt3Vq+GZZ9yNgbmZdV97zTLrJoiUChp2ecqYFHDkkdCvH6xYAS+9BH/+Cd26wemnw6RJsG9fvGtYrKVU0LDLU8akkJIloUcPWLIEXn/dnYn07u3Sso8bB3v2xLuGxVJKBQ1jTAoqUQK6dj2UWfdvf4MBA9z9H6NGFcvMuvFkQcMYkxz8M+vWqgW33losM+vGU0oFDWvTMKYYyM2sO3s2zJsHZ55Z7DLrxlNKBQ1r0zCmmDnrLHjvPViwwGXWfeghd+Zxxx3w00/xrl1KSqmgYYwppho3dpl1lyxxmXUfe8wFj4ED4ccf4127lGJBwxiTOk4/3XXTXb7cZdMdP941mPftCytXxrt2KcGChjEm9dSoAc8+624UHDAAXn3VMutGSUoFDWsIN8Yc5qSTYPRol6Lk9tvhnXegXj13CcsyYRdISgUNawg3xgRUoYIbCMo/s+4FF1hm3XxKqaBhjDEh+WfWXbDAZdZt1Qo++siSI0bAgoYxpvjxzaz7+OOwahW0awfNmrlLWBY8grKgYYwpvo46CgYNOpRZd+tW6NQJGjWCqVMts24AFjSMMcY3s+6LL8LevXDFFVCnjmXW9ZNSQcN6TxljCqVkSXd/x+LFLrNuevrhmXX37o13DeMupYKG9Z4yxkRFWprLrLtwoWvjqFDB3e9RvbprAynGmXVTKmgYY0xUiUDHjvDll653Va1acMstLkXJI4/Azp3xrmGRs6BhjDHhiECbNi6z7uefu3s8/vUvl1n3vvtg27Z417DIWNAwxpj8aNEC3n/f3eNx3nkwZIgLHv/8Z7HIrGtBwxhjCqJxY3jzTddo3qULPPooVKsGN96Y0pl1LWgYY0xh1KkDL7/sMuv26OHu9zjlFLj2WnfTYIqxoGGMMdGQm1l31Sro3x9eecU1nF91lRvnI0VY0DDGmGg6+WR48klYuxZuuw3efhvq1oVLL3XtIEkupYKG3dxnjEkYf/sb/Oc/Lr/VfffBJ59AVpbLrDtvXrxrV2ApFTTs5j5jTMI54QR48EGXWfeRR9zZxtlnQ+vWSZlZN6WChjHGJKxjjoHBgw9l1l2xwmXWbd48qTLrWtAwxpiilJtZd80al89q8+akyqxrQcMYY+LhyCPh+uvdGcekSbBnz6HMui++mLCZdS1oGGNMPJUqBVdf7brlTp3qgkmvXi6z7jPPJFxmXQsaxhiTCNLS4PLLYdGiQ5l1+/c/lFn3jz/iXUPAgoYxxiQW/8y6p56aUJl1LWgYY0wiys2s+8knLrNuVlZCZNa1oGGMMYkuN7Nudra7vyOOmXUtaBhjTLI44wx46y347jvo3DkumXUtaBhjTLKpW9clRFy+3CVEHDfuUGbd7dtjuumEDxoiUl1EnhORN+JdF2OMSSg1asCECbB6tetpNWcOHH10TDcZ06AhIs+LyBYRWew3vYOILBeRVSIyOFQZqrpGVfvGsp7GGJPUcjPrLlvm7vOIoZIxLR0mAk8BL+ZOEJE0YAzQDlgPzBeRt4E04BG/9a9R1S0xrqMxxqSGUqVivomYBg1V/UxEMv0mnwmsUtU1ACIyBeisqo8AHWNZH2OMMYUTjzaNysA6n+frvWkBicgJIjIOaCQid4VYrp+IZItI9tatW6NXW2OMMXlifXkqEAkwLWhOYFXdBvQPV6iqjgfGA2RlZSVHjmFjjEky8TjTWA+c5PO8CrAxGgXbyH3GGBNb8Qga84GaIlJNRI4AugFvR6NgG7nPGGNiK9ZdbicDXwK1RGS9iPRV1f3AQGAmsAyYqqpLorQ9O9MwxpgYEk2SIQbzIysrS7Ozs+NdDWOMSRoiskBVs8Iul4pBQ0S2Aj/4TDoWCHT6EWj6icDPMapafgSrc1GXl5/1Ilk23DL52Vehptt+LPh6hd2PBZlnn8XorleQfVhVVcuFLVlVU/4BjI90OpAd7/qGqnNRl5ef9SJZNtwy+dlXYabbfozTfizIPPssJtY+DPVI+NxTUfJOPqcngmjXraDl5We9SJYNt0x+91Ui70MonvuxIPMSeT8Wx30YVEpenioMEcnWCK7rmcRm+zH52T5MTMXlTCM/xse7AiYqbD8mP9uHCcjONIwxxkTMzjSMMcZEzIKGMcaYiFnQMMYYEzELGiGIyNEiMklEnhWRq+JdH1MwNmRwahCRLt5ncYaInB/v+hRXxS5o5HMI2kuBN1T1OqBTkVfWBJWf/ag2ZHDCyud+nO59FnsDV8ShuoZiGDRwQ9B28J3gMwTtBcDpQHcROR2Xtj13wKgDRVhHE95EIt+PJnFNJP/78R5vvomDYhc0VPUzYLvf5LwhaFX1T2AK0Bk39kcVb5li914lsnzuR5Og8rMfxRkOfKCqXxd1XY1jX4ROsCFo3wIuE5GnSew0B8YJuB8jHTLYJIxgn8cbgbZAVxEJO5qniY14DPeaiAIOQauqvwN9iroypsCC7ceIhgw2CSPYfhwNjC7qypjD2ZmGE7MhaE2Rsv2YGmw/JjALGk7MhqA1Rcr2Y2qw/ZjAil3QKOohaE1s2H5MDbYfk48lLDTGGBOxYnemYYwxpuAsaBhjjImYBQ1jjDERs6BhjDEmYhY0jDHGRMyChjHGmIhZ0DBJT0QOiMgiEVkiIt+IyK0iklLHtog0FJELC7BeZm7acRHJEpHR3v+tROSsaNfTpD7LPWVSwW5VbQggIuWBV4FjgfsLW7CIpKlqIqTFbwhkAe/7zxCRkt4NcSGpajaQ7T1tBewCvohiHU0xkFK/xoxR1S1AP2Cgl0o7TURGiMh8EflWRK4HEJESIjLWOzt5V0TeF5Gu3rwcEblPRD4HLheRU0TkvyKyQETmishp3nLlRORNr+z5ItLCvz7e9keKyHfe9m/0pp8hIp96Zc4UkYre9DkiMlxEvhKRFSLS0kul8W/gCu+M6goReUBExovIh8CL3hnFXBH52nv85SzCO7t4V0QycQkcb/HKaykia0WklLfcMd57UCra+8ckPzvTMClHVdd4l6fK48bT+FVVm4jIkcA874v2DCATqOcttwx43qeYPap6NoCIfAz0V9WVItIUGAucBzwBjFLVz0XkZFzai9p+1ekHVAMaqep+ETne+zJ+EuisqltF5ApgKHCNt05JVT3Tuxx1v6q2FZH7gCxVHejV6QHvNZytqrtF5CignaruEZGawGTcmUmg9yfHSxW/S1VHeuXNAS4CpuNyPb2pqvsiftNNsWFBw6Sq3PTa5wP1c88icJetagJnA6+r6kHgJxH5xG/91wBEpDRwFvC6SF7G7iO9v22B032mHyMiZVT1N59y2gLjci8fqep2EakL1AVmeeumAZt81nnL+7sAF9iCeVtVd3v/lwKeEpGGuFEmTw2xXiATgH/igkYf4Lp8rm+KCQsaJuWISHXcF+cWXPC4UVVn+i1zUZhifvf+lgB25LaZ+CkBNPf54g5YHcA/wZsAS1S1eZB19np/DxD6M/q7z/+3AJuBBl699oRY7y9UdZ53ietcIE1VF4ddyRRL1qZhUoqIlAPGAU+py8Y5Exjgc73+VBE5GvgcNypjCRGpgGsY/gtV3QmsFZHLvfVFRBp4sz/EZWPN3XagwPIh0F9ESnrLHA8sB8qJSHNvWikRqRPmpf0GlAkx/1hgk3fm1BN39pLf8l7EXdZ6Icy6phizoGFSQUZul1vgI9wX9YPevAnAUuBrr+vpM7hf72/iBvvJnfY/4Ncg5V8F9BWRb4AlHBp3/CYgy2vgXkrg0QEnAD8C33rrX+mNe90VGO5NW4S7BBbKJ7hLYYu8NhB/Y4FeIvJ/uEtTvwdYxtc7wCW5DeHetFeA43CBw5iALDW6KbZEpLSq7hKRE4CvgBaq+lO86xUvXrtPZ1XtGe+6mMRlbRqmOHtXRMoCRwBDinnAeBK4AMj3DYSmeLEzDWOMMRGzNg1jjDERs6BhjDEmYhY0jDHGRMyChjHGmIhZ0DDGGBMxCxrGGGMi9v/xg6qXDvrCtwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "# Assuming you already have the graph 'G'\n",
    "\n",
    "# Calculate degree centrality for each node\n",
    "degree_centrality = nx.degree_centrality(G)\n",
    "degrees = [int(round(dc * (len(G)-1))) for dc in degree_centrality.values()]\n",
    "\n",
    "# Count the number of nodes for each degree centrality\n",
    "degree_counts = Counter(degrees)\n",
    "\n",
    "# Define a function for the power-law fitting in log-log space\n",
    "def linear_func(x, a, b):\n",
    "    return a + b * x\n",
    "\n",
    "# Fit the power-law function to the log-transformed data\n",
    "x = np.array(sorted(list(degree_counts.keys())))\n",
    "y = np.array([degree_counts[d] for d in x])\n",
    "\n",
    "# Filter out zero values in x and y\n",
    "non_zero_indices = (x > 0) & (y > 0)\n",
    "x_filtered = x[non_zero_indices]\n",
    "y_filtered = y[non_zero_indices]\n",
    "\n",
    "log_x = np.log(x_filtered)\n",
    "log_y = np.log(y_filtered)\n",
    "\n",
    "# Use Levenberg-Marquardt algorithm for curve fitting\n",
    "popt, _ = curve_fit(linear_func, log_x, log_y, method='lm')\n",
    "\n",
    "# Calculate the fitted power-law function using the log-transformed parameters\n",
    "fitted_power_law = lambda x: np.exp(popt[0]) * np.power(x, popt[1])\n",
    "\n",
    "# Plot the scatter plot and the fitted power-law function\n",
    "plt.scatter(x, y, label='Data')\n",
    "plt.plot(x, fitted_power_law(x), 'r-', label=f'Fitted power-law: y = {np.exp(popt[0]):.2f}x^({popt[1]:.2f})')\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.xlabel('Degree centrality')\n",
    "plt.ylabel('Number of nodes')\n",
    "plt.legend()\n",
    "plt.title('Number of nodes as a function of their degree centrality')\n",
    "\n",
    "# Save the output as a PNG file\n",
    "plt.savefig('2000g.png', dpi=150)\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
